### ISSUE TITLE
Post per page limit in the themes

### ISSUE THREAD

**Issue by dingo-d** [WPTRT/WPThemeReview#147]
We should make an agreement on if we should have some fixed post per page limit, except disallowing `-1`, and handling `nopaging` in queries.

A discussion in the WPTRT would be needed to reach a conclusion.

**Comment by justintadlock** [comment]
Generally speaking, we've said that themes should not touch the core queries (e.g., blog posts page, author archive, etc.).  That falls under plugin territory.  Custom page templates and a front page template are the obvious exception to that.

**Comment by justintadlock** [comment]
With the above said, there are some special cases.  I could theoretically see some type of directory theme that lists off everything and alters the main query in other ways.  Of course, that'd fall under a special case and wouldn't be a part of a standard.

**Comment by dingo-d** [comment]
Yeah, you could hook to `pre_get_posts` and alter any query in the theme (wouldn't see why that couldn't be in the theme tbh).
Some sensible number should be agreed upon, so that we can add it in the ruleset, and in the handbook

**Comment by justintadlock** [comment]
A sensible number is "don't touch it unless you absolutely must because this is generally plugin territory".  That was already discussed long ago.  Putting an arbitrary number up as a limit doesn't help when it's not allowed and would never address all the use cases for when there's an exception.

Frankly, this topic seems like we're looking for a solution to a problem that doesn't really exist.  

Are we suddenly getting themes that are querying 110 posts but 100 seems like a nicer cut-off point?  What's the reason for this ticket?

**Comment by dingo-d** [comment]
The reason for this ticket is the existing [posts per page sniff](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/WordPress/Sniffs/VIP/PostsPerPageSniff.php) that we need to make a decision what to do with it - if we allow it in the TRT CS then some limit needs to be set. 

I'd rather include it to prevent cases where people query the entire table (setting `posts_per_page => -1`). We could just include that case, and leave the number to be whatever (10, 50, 10000)...

**Comment by jrfnl** [comment]
@dingo-d Just FYI: As an alternative, a sniff could be added to completely forbid touching `posts_per_page` from within a theme.

**Comment by justintadlock** [comment]
We don't need such a thing.  If there's an existing sniff, simply ignore it for the purposes of TRT.  

There are legit cases to set `'posts_per_page' => -1`.  Just as a quick example off the top of my head, a theme might build a custom page template that creates an archive of pages or some other post type.

**Comment by dingo-d** [comment]
> There are legit cases to set 'posts_per_page' => -1

Let's agree to disagree. Unless you have a blog with over 100k posts in it, setting posts_per_page to `-1` has a huge negative effect on the query speed. I did the test on one of my projects where I had huge number of posts and had to search through all of them, and even setting `posts_per_page` to `50000` was better than setting it to -1.
Sure, if you have a huge number of posts you could run into a problem of passing this limit. But these are really big edge cases.

Anyways, I'd like to see what other members think about this. @pattonwebz could bring this up on one of the meetings, and if the people are for ignoring this rule, we can ignore it :)

**Comment by grappler** [comment]
I agree that in most cases `-1` is not needed. Though if you have `50000` posts then you will not likely be using a theme from w.org and will have your own team who will tell you to use `-1` is not a good idea. Related: https://10up.github.io/Engineering-Best-Practices/php/

**Comment by justintadlock** [comment]
Team decided it wants this check but to keep it as a warning rather than an error for some edge cases.

Meeting: https://wordpress.slack.com/archives/C02RP4Y3K/p1528822802001090

**Comment by dingo-d** [comment]
If I'm not mistaken this would just require the change in the `ruleset.xml` so

```xml
<rule ref="WordPress.WP.PostsPerPage">
  <type>warning</type>
</rule>
```

If that's correct it's a minor change and can be added in 0.2.0 release 🙂

**Comment by jrfnl** [comment]
@dingo-d The `WordPress.WP.PostsPerPage` sniff checks for a *high* (> 100) limit, not for `-1`. That part was contained in the `VIP` sniff which has been removed, though there is discussion in WPCS about bringing it back.
See: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/1638

**Comment by jrfnl** [comment]
Closing as this has been addressed via PR #218.
